core/src/main/java/com/datastrato/gravitino/lock/TreeLock.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.lock;
+
+import java.util.List;
+import java.util.Stack;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * TreeLock is a lock that manages the lock process of the resource path. It will lock the whole
+ * path from root to the resource path.
+ *
+ * <p>Assuming we need to alter the table `metalake.catalog.db1.table1`, the lock manager will lock
+ * the following
+ *
+ * <pre>
+ *   /                                    readLock
+ *   /metalake                            readLock
+ *   /metalake/catalog                    readLock
+ *   /metalake/catalog/db1                readLock
+ *   /metalake/catalog/db/table1          writeLock
+ * </pre>
+ *
+ * If the lock manager fails to lock the resource path, it will release all the locks that have been
+ * locked in the inverse sequences it locks the resource path.
+ */
+public class TreeLock {
+  public static final Logger LOG = LoggerFactory.getLogger(TreeLock.class);
+
+  private final List<TreeLockNode> lockNodes;
+  private final Stack<TreeLockNode> heldLocks = new Stack<>();
+  private LockType lockType;
+
+  public TreeLock(List<TreeLockNode> lockNodes) {
+    this.lockNodes = lockNodes;
+  }
+
+  /**
+   * Lock the tree lock with the given lock type.
+   *
+   * @param lockType The lock type to lock the tree lock.
+   */
+  public void lock(LockType lockType) {
+    this.lockType = lockType;
+
+    int length = lockNodes.size();
+    for (int i = 0; i < length; i++) {
+      TreeLockNode treeLockNode = lockNodes.get(i);
```

Comment:
If the lock remains locked for a long time, we should throw an exception to the user. The exception message should include the lock name, the ID of the thread that holds the lock, and the duration of the lock. This can prevent all threads from hanging on the lock and assist us in debugging the problems